### PR TITLE
Improve handling of cluster scale down

### PR DESCRIPTION
**Pull request by bjosv**
Make sure that `ctx_get_by_node()` always returns a viable connection context,
i.e. it will return a context or NULL if a connection attempt failed.
    
If `redis_cluster_command_execute()` fails to get a viable context it will attempt to find another node to send the message to.
This will will usually trigger a cluster topology update.

Added a testcase for the scenario where a node that has not previously been accessed is removed.

TODO:
- How to handle OOM errors, maybe a OOM error should bail instead of trying to create a context for the next node.
 The triggered changes in the OOM test..

### CODE REVIEW

**Comment by bjosv** [comment]
CC @Moo64c

File: hircluster.c [hunk 1]
```diff
@@ -1263,12 +1263,11 @@ static int cluster_update_route_by_addr(redisClusterContext *cc, const char *ip,
         goto error;
     }
 
-    if (cc->connect_timeout) {
-        c = redisConnectWithTimeout(ip, port, *cc->connect_timeout);
-    } else {
-        c = redisConnect(ip, port);
-    }
+    redisOptions options = {0};
```
Comment (bjosv) [inline hircluster.c:1266]: Maybe this should have been put in a separate PR

File: hircluster.c [hunk 2]
```diff
@@ -2008,42 +2007,29 @@ int redisClusterConnect2(redisClusterContext *cc) {
     return _redisClusterConnect2(cc);
 }
 
+/* Get a viable connection context to a given cluster node.
+ * Will connect to the node, setup TLS and authenticate when required.
+ * If an existing context exist but with errors it will be re-created.
+ */
```
Comment (zuiderkwast) [inline hircluster.c:2013]: So this function never returns a connection in an error state? It aways returns NULL in that case?

File: hircluster.c [hunk 3]
```diff
@@ -2398,8 +2381,6 @@ static void *redis_cluster_command_execute(redisClusterContext *cc,
 
     c = ctx_get_by_node(cc, node);
     if (c == NULL) {
-        return NULL;
-    } else if (c->err) {
         node = node_get_which_connected(cc);
```
Comment (zuiderkwast) [inline hircluster.c:2384]: I'm thinking maybe there is a point to distinguish between initial connect failure and reconnect failure. Maybe that was the idea of the old code. I hope you can prove me wrong. :-)

What if connection fails the first time when trying to connect to a node? We will pick a different node, get a MOVED redirect, update the slot mapping, then try to connect the failing node again and loop(?).

The idea of doing it only at failed *re*connect is that hopefully an updated slot mapping will give us the right node. It may still loop though.

**Review by zuiderkwast — COMMENTED** [review]
Beautiful cleanup. Let's reason about what it implies e.g. for failing initial connect to a node and if that's good or not.

File: hircluster.c (hunk 1 above)
Comment (zuiderkwast) [reply hircluster.c:1266]: Doesn't matter, but we should fix the build before merging. Add an explicit cast or so, I guess.

Comment (bjosv) [reply hircluster.c:1266]: I thought this was possible from c99, maybe its some kind of gcc extension not available in early clang.
Anyway its fixable.

File: hircluster.c (hunk 2 above)
Comment (bjosv) [reply hircluster.c:2013]: Yes, the idea was to make sure its always returns a connection context without any known errors.
NULL if a connect/reconnect attempt fails

Comment (zuiderkwast) [reply hircluster.c:2013]: OK, good. I think the comment can make that even more clear.

File: hircluster.c (hunk 1 above)
Comment (zuiderkwast) [reply hircluster.c:1266]: Probably the old clang didn't follow c99 correctly.

Redis itself compiles with the following warning flags: `-Wall -W -Wno-missing-field-initializers`.

File: hircluster.c (hunk 3 above)
Comment (bjosv) [reply hircluster.c:2384]: Looks like this has been there from day 1 in hiredis-vip, so git blame didn't help :)
If its a scaledown then the node can be gone before we attempt to connect to for the first time, then we need to get a new topology view.
But for a problem like a routing problem, then we will do max-retry-count number of attempts (default 5) and getting the same topology each time, a bit wasteful maybe?
There is currently no check that a topology view update via `cluster_update_route()` gives the same view as we already know, maybe that would help for that case..

File: hircluster.c (hunk 1 above)
Comment (bjosv) [reply hircluster.c:1266]: Ah interesting, looked at hiredis which also have it:
`-Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers`

but its not defined in its CMakeLists.txt weirdly enough.

**Comment by zuiderkwast** [comment]
@Moo64c Björn is on vacation. He'll be back in 3 weeks IIRC. I believe we can include this PR in the next release, soon after he comes back from vacation.

**Comment by Moo64c** [comment]
@bjosv @zuiderkwast  Hi,

1. We're considering using this branch as a temporary patch for an upcoming release. Do you have any relevant suggestions (not ready, problematic, need an additional fix etc.)? 
2. We will likely have some time to assist with the PR next month, if required - are there any things left to cover to help this PR make it into the next version?

Thanks

**Comment by bjosv** [comment]
Hi @Moo64c I will try to restart this work soon, but we are happy to get so help.

If I remember correctly there were scenarios that needed to be thought though, like how to handle temporary tcp-disconnects vs. vanished redis nodes as in your case. When to fetch a new cluster topology while making sure we don't do that all to often.
If the patch works in your testing I believe that's a good start, I have not yet battle-tested it myself and would like to add some tests before delivering it.

**Comment by Moo64c** [comment]
@bjosv Great!

Considering temporary disconnects :
- the driver has a definition for connection timeout - can it be used to determine we waited long enough for a remapping? 
- Maybe an additional configuration option is in order?
-  If I remember correctly there's already tracking on how many times the driver failed to contact a node, maybe use that?

In any case we will test to the best of our abilities but since the scenario is very specific it is somewhat likely to be missed in testing.

**Comment by lior-parsi** [comment]
Hi,
is there any plan to issue this fix?

**Comment by zuiderkwast** [comment]
> Make sure that `ctx_get_by_node()` always returns a viable connection context, i.e. it will return a context or NULL if a connection attempt failed.

I don't get this sentence. If it can return NULL, it does not always return a viable connection context.

> If `redis_cluster_command_execute()` fails to get a viable context it will attempt to find another node to send the message to. This will will usually trigger a cluster topology update.

This seems to be the main change in this PR. It should be the first and foremost sentence. :-)

Btw, what if `redis_cluster_command_execute()` is used for some admin command for a specific node? We can't simply send that to another node... Maybe it's better that a failed connection attempt simply triggers a slotmap update but doesn't resend the command to a random node?

File: hircluster.c (hunk 3 above)
Comment (zuiderkwast) [reply hircluster.c:2384]: The whole random node thing is a quite dangerous. What if we want to send an admin command to a specific node (using ToNode) which is down. Then it is sent to a random node instead(!).

I think we can update the slotmap, then try again to fetch the node responsible for a certain key, but not send a command to a random node.

**Comment by bjosv** [comment]
> > Make sure that `ctx_get_by_node()` always returns a viable connection context, i.e. it will return a context or NULL if a connection attempt failed.
> 
> I don't get this sentence. If it can return NULL, it does not always return a viable connection context.

If I remember correctly the idea to have `ctx_get_by_node()` return healthy hiredis context if any.
 
> > If `redis_cluster_command_execute()` fails to get a viable context it will attempt to find another node to send the message to. This will will usually trigger a cluster topology update.
> 
> This seems to be the main change in this PR. It should be the first and foremost sentence. :-)
> 
Maybe, but `ctx_get_by_node()` was used in other places as well with similar handling as to which we changed.
I guess we have changed it alot now, so I cant really compare. This PR would problably not fit anymore.. maybe I should close it.

> Btw, what if `redis_cluster_command_execute()` is used for some admin command for a specific node?

`redis_cluster_command_execute()` will only be used for slot routed commands I believe, which will attempt to find the key.
The `ToNode()` API should be used to send to specific nodes (which should use `redis_cluster_command_execute()`  internally?)

**Comment by zuiderkwast** [comment]
> > > Make sure that `ctx_get_by_node()` always returns a viable connection context, i.e. it will return a context or NULL if a connection attempt failed.
> > 
> > I don't get this sentence. If it can return NULL, it does not always return a viable connection context.
> 
> If I remember correctly the idea to have `ctx_get_by_node()` return healthy hiredis context if any.

Ah, so "if any" is the missing part in the sentence. It's something about the grammar which confused me. First it says that it always returns a viable context. Then it says it returns a context or NULL. It's confusing in my head since the second part contradicts the first part.

You actually mean that it never returns a context with an error?

**Comment by bjosv** [comment]
Solved in #182 and the testcase added via #183.
